crates/circuit/src/circuit_data.rs

```diff
@@ -2414,6 +2766,14 @@ impl CircuitData {
         self.cargs_interner().get(index)
     }
 
+    /// Internal method for assigning parameters.
+    ///
+    /// Note that currently if any [ParameterUse] identifies a basic block,
+    /// the block will be replaced for all instructions in the circuit, even if
+    /// they aren't included in the use set. This is because we would otherwise
+    /// have to keep a mapping of [Block] to all instructions that use it and
+    /// add/create an additional block if only some uses are replaced. This is
+    /// not a use-case that we currently have.
```

Comment:
This feels weird, and like the `ParameterTable` isn't tracking things entirely correctly any more.  It's not possible for a `Parameter` to be used by one reference to a basic block, but not by a different reference - either the basic block has the parameter or it doesn't.  Feels like the `ParameterTable` should be able to encode that, most likely by `ParamUsage` gaining a new variant (e.g. `Block`, to go with `GlobalPhase`, `Instruction`), so it's decoupled from the per-instruction one.

We'd have to take care that the block-mapping functions updated the parameter table in that case (which tbh, I think we still need to take care of?).